pkg/operator/ceph/pool/radosnamespace/controller.go

```diff
@@ -0,0 +1,316 @@
+		}
+
+		// Return and do not requeue. Successful deletion.
+		return reconcile.Result{}, nil
+	}
+
+	if !cephCluster.Spec.External.Enable {
+		// Build the NamespacedName to fetch the CephBlockPool and make sure it exists, if not we cannot
+		// create the rados namespace
+		cephBlockPool := &cephv1.CephBlockPool{}
+		cephBlockPoolRadosNamespacedName := types.NamespacedName{Name: cephBlockPoolRadosNamespace.Spec.BlockPoolName, Namespace: request.Namespace}
+		err = r.client.Get(r.opManagerContext, cephBlockPoolRadosNamespacedName, cephBlockPool)
+		if err != nil {
+			if kerrors.IsNotFound(err) {
+				return reconcile.Result{}, errors.Wrapf(err, "failed to fetch ceph blockpool %q, cannot create rados namespace %q", cephBlockPoolRadosNamespace.Spec.BlockPoolName, cephBlockPoolRadosNamespace.Name)
+			}
+			// Error reading the object - requeue the request.
+			return reconcile.Result{}, errors.Wrap(err, "failed to get cephBlockPoolRadosNamespace")
+		}
+
+		// If the cephBlockPool is not ready to accept commands, we should wait for it to be ready
+		if cephBlockPool.Status.Phase != cephv1.ConditionReady {
+			// We know the CR is present so it should a matter of second for it to become ready
+			return reconcile.Result{Requeue: true, RequeueAfter: 10 * time.Second}, errors.Wrapf(err, "failed to fetch ceph blockpool %q, cannot create rados namespace %q", cephBlockPoolRadosNamespace.Spec.BlockPoolName, cephBlockPoolRadosNamespace.Name)
+		}
+	}
+	if cephCluster.Spec.External.Enable {
+		logger.Debugf("external rados namespace %q creation is not supported, create it manually, the controller will assume it's there", namespacedName)
+	} else {
+		// Create or Update rados namespace
+		err = r.createOrUpdateRadosNamespace(cephBlockPoolRadosNamespace)
+		if err != nil {
+			if strings.Contains(err.Error(), opcontroller.UninitializedCephConfigError) {
+				logger.Info(opcontroller.OperatorNotInitializedMessage)
+				return opcontroller.WaitForRequeueIfOperatorNotInitialized, nil
+			}
+			r.updateStatus(r.client, request.NamespacedName, cephv1.ConditionFailure)
+			return reconcile.Result{}, errors.Wrapf(err, "failed to create or update ceph pool rados namespace %q", cephBlockPoolRadosNamespace.Name)
+		}
+	}
+	// Update CSI config map
+	// If the mon endpoints change, the mon health check go routine will take care of updating the
+	// config map, so no special care is needed in this controller
+	csiClusterConfigEntry := csi.CsiClusterConfigEntry{
+		Monitors:       csi.MonEndpoints(r.clusterInfo.Monitors),
+		RadosNamespace: cephBlockPoolRadosNamespace.Name,
+	}
+	err = csi.SaveClusterConfig(r.context.Clientset, buildClusterID(cephBlockPoolRadosNamespace), r.clusterInfo, &csiClusterConfigEntry)
+	if err != nil {
+		return reconcile.Result{}, errors.Wrap(err, "failed to save cluster config")
+	}
+	// Success! Let's update the status
+	if cephCluster.Spec.External.Enable {
+		r.updateStatus(r.client, request.NamespacedName, cephv1.ConditionConnected)
```

Comment:
Why do we have a "connected" status for the rados namespace? The cluster CR tracks if there is a connection, but this CR only creates/deletes rados namespaces, right? Seems like it's either created or not, rather than connected.

Reply:
And according to the check above, creating a rados namespace from an external cluster isn't supported. Shouldn't that check just return an error and fail the reconcile? Or what is the purpose in reconciling the rados namespace for an external cluster?

Reply:
>Why do we have a "connected" status for the rados namespace? The cluster CR tracks if there is a connection, but this CR only creates/deletes rados namespaces, right? Seems like it's either created or not, rather than connected.

This is the same as the https://github.com/rook/rook/blob/master/pkg/operator/ceph/file/subvolumegroup/controller.go#L265-L272  nothing new in this one. i have seen in other places also `cephv1.ConditionType` used for status.Phase, just reusing the same here also.

>Or what is the purpose in reconciling the rados namespace for an external cluster?

This is requested to generate the clusterID which is required for the storageclass creation, the whole idea of supporting external cluster got added in https://github.com/rook/rook/pull/9694

Reply:
@leseb What is the need to support subvolume groups or rados namespaces as "connected" status in external mode? It seems like nothing is reconciled in external mode, not sure what external mode is doing here. "connected" just seems like a status for the cluster CR, not other CRs like this.

Reply:
The external condition is a bit of a hack/noop to not perform some commands and expect the producer cluster (external) to perform the operation instead.
I tend to agree that we could have avoided the "Connected" state for subvolumegrp CRD and perhaps this one but we need to decide if we keep it for this new controller or change both :)

Reply:
Just to be clear, the controllers don't need to do anything for external clusters for the rados namespaces and subvolume groups, right? In that case I'd vote that we don't support them in external mode and just error out if they are created in an external cluster.

Reply:
No they do need to do something on external mode, which is updating the ceph-csi configmap to input the subvolumegroup/namespace crd name cpeh-csi should use.

Reply:
Ok I see now where the configmap is updated. I think what would help is just a little code refactor like suggested above so it's more obvious.